src/cpp/cpp_typecheck_type.cpp

```diff
@@ -126,7 +127,10 @@ void cpp_typecheckt::typecheck_type(typet &type)
     exprt &size_expr=to_array_type(type).size();
 
     if(size_expr.is_not_nil())
+    {
       typecheck_expr(size_expr);
+      simplify(size_expr, *this);
+    }
```

Comment:
It's quite a while since I built that patch, but I suspect that is necessary to make some of the C++ regression tests pass. Also it is the same that the C front-end does.